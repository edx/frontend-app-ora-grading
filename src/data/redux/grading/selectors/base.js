import { createSelector } from 'reselect';

import { StrictDict } from 'utils';
import * as module from './base';

export const rootSelector = ({ grading }) => grading;
export const mkSimpleSelector = (key) => createSelector(
  [module.rootSelector],
  (grading) => grading[key],
);
export const rootKeys = StrictDict({
  selected: 'selected',
  activeIndex: 'activeIndex',
  current: 'current',
  gradeData: 'gradeData',
  gradingData: 'gradingData',
});

export const simpleSelectors = {
  selected: mkSimpleSelector(rootKeys.selected),
  activeIndex: mkSimpleSelector(rootKeys.activeIndex),
  current: mkSimpleSelector(rootKeys.current),
  gradeData: mkSimpleSelector(rootKeys.gradeData),
  gradingData: mkSimpleSelector(rootKeys.gradingData),
};

/**
 * returns the length of the list of selected submissions
 * @return {number} selected submission list length
 */
export const selectionLength = createSelector(
  [module.simpleSelectors.selected],
  (selected) => selected.length,
);

export default {
  ...simpleSelectors,
  selectionLength,
};
