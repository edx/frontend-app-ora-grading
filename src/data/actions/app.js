import { StrictDict } from 'utils';
// import { createActionFactory } from './utils';

// export const dataKey = 'app';
// const createAction = createActionFactory(dataKey);

export default StrictDict({ });
