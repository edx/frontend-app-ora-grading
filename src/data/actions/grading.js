import { StrictDict } from 'utils';
import { createActionFactory } from './utils';

export const dataKey = 'grading';
const createAction = createActionFactory(dataKey);

/**
 * Load the first of the selected submission list for review, and initializes
 * the review pane to the first index.
 * @param {obj} submission data for the review/grading view
 *   {
 *     {obj} response - api response data
 *     {obj} gradeData - api grade data
 *     {str} status - api grade status
 *   }
 */
const loadSubmission = createAction('loadSubmission');

/**
 * Pre-load just the static info about the "next" submission in the review queue.
 * Load submission and the learner's response.
 * @param {obj} submission ({ response })
 */
const preloadNext = createAction('preloadNext');

/**
 * Pre-load just the static info about the "previous" submission in the review queue.
 * Load submission and the learner's response.
 * @param {obj} submission ({ response })
 */
const preloadPrev = createAction('preloadPrev');

/**
 * Load the "next" submission in the selected queue as the current selection, load its current
 * status and grade data, and update prev/next accordingly.
 * @param {obj} { status, gradeData }
 */
const loadNext = createAction('loadNext');

/**
 * Load the "prev" submission in the selected queue as the current selection, load its current
 * status and grade data, and update prev/next accordingly.
 * @param {obj} { status, gradeData }
 */
const loadPrev = createAction('loadPrev');

/**
 * Load the selected submissions, storing their static data in an ordered array and setting the starting
 * index at the beginning of the list.
 * @param {obj[]} selection - ordered array of submission static data for all selected submissions
 */
const updateSelection = createAction('updateSelection');

// TODO: implement/design data workflow
const rubric = StrictDict({
  /*
   * update the local version of the rubric-level comment
   * @param {string} comment
   */
  updateComment: createAction('rubric/comment'),
  /*
   * update the local version of points for the given criterion
   * @param {number} index
   * @param {number} points
   */
  updateCriterionPoints: createAction('rubric/criterionPoints'),
  /*
   * update the local version of comment for the given criterion
   * @param {number} index
   * @param {string} comments
   */
  updateCriterionComment: createAction('rubric/criterionComment'),
});

export default StrictDict({
  loadSubmission,
  preloadNext,
  preloadPrev,
  loadNext,
  loadPrev,
  updateSelection,
  rubric,
});
