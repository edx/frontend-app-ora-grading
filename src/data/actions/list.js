import { StrictDict } from 'utils';
import { createActionFactory } from './utils';

export const dataKey = 'list';
const createAction = createActionFactory(dataKey);

export default StrictDict({
  load: createAction('load'),
  updateSelection: createAction('updateSelection'),
});
