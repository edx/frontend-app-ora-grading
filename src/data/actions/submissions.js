import { StrictDict } from 'utils';
import { createActionFactory } from './utils';

export const dataKey = 'submissions';
const createAction = createActionFactory(dataKey);

/**
 * Load the basic list-level submission data, keyed by submission id
 * @param {obj} submissionListData
 */
const loadList = createAction('loadList');

export default StrictDict({
  loadList,
});
