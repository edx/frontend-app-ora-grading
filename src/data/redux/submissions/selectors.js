import _ from 'lodash';
import { createSelector } from 'reselect';

import { StrictDict } from 'utils';
import { lockStatuses } from 'data/services/lms/constants';

export const simpleSelectors = {
  allSubmissions: state => state.submissions.allSubmissions,
};

/**
 * Returns the submission list in default order for the table.
 */
export const listData = createSelector(
  [simpleSelectors.allSubmissions],
  (allSubmissions) => {
    const submissionIds = Object.keys(allSubmissions);
    const submissionList = submissionIds.map(id => {
      const { gradeStatus, lockStatus, ...rest } = allSubmissions[id];
      const gradingStatus = (lockStatus === lockStatuses.unlocked ? gradeStatus : lockStatus);
      return { gradingStatus, ...rest };
    });
    return _.sortBy(
      submissionList,
      ['submissionDate'],
    );
  },
);

export default StrictDict({
  ...simpleSelectors,
  listData,
});
