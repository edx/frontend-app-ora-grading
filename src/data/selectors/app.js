import { StrictDict } from 'utils';

export const simpleSelectors = {
  showReview: state => state.app.showReview,
  showRubric: state => state.app.showRubric,
  grading: state => state.app.grading,
  courseMetadata: state => state.app.courseMetadata,
  oraName: state => state.app.oraMetadata.name,
  oraPrompt: state => state.app.oraMetadata.prompt,
  oraTypes: state => state.app.oraMetadata.type,
};

export default StrictDict({
  ...simpleSelectors,
});
