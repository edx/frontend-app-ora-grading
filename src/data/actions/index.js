import { StrictDict } from 'utils';

import app from './app';
import grading from './grading';
import submissions from './submissions';

export default StrictDict({
  app,
  grading,
  submissions,
});
