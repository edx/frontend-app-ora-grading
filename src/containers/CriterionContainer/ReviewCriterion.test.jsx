import React from 'react';
import { shallow } from 'enzyme';

import actions from 'data/actions';
import selectors from 'data/selectors';
import { ReviewCriterion, mapStateToProps } from './ReviewCriterion';

jest.mock('@edx/paragon', () => ({
  Form: {
    Label: () => 'Form.Label',
  },
  FormControlFeedback: () => 'FormControlFeedback',
}));

jest.mock('data/selectors', () => ({
  __esModule: true,
  default: {
    app: {
      rubric: {
        criterionConfig: jest.fn((...args) => ({
          rubricCriterionConfig: args,
        })),
      },
    },
    grading: {
      selected: {
        criterionGradeData: jest.fn((...args) => ({
          selectedCriterionGradeData: args,
        })),
      },
    },
  },
}));

describe('Review Crition Container', () => {
  const props = {
    orderNum: 1,
    config: {
      prompt: 'prompt',
      name: 'random name',
      feedback: 'feedback mock',
      options: [
        {
          explanation: 'explaination',
          feedback: 'option feedback',
          label: 'this label',
          name: 'option name',
          points: 1,
        },
        {
          explanation: 'explaination 2',
          feedback: 'option feedback 2',
          label: 'this label 2',
          name: 'option name 2',
          points: 2,
        },
      ],
    },
    data: {
      selectedOption: 'selected option',
      feedback: 'data feedback',
    },
  };

  let el;
  beforeEach(() => {
    el = shallow(<ReviewCriterion {...props} />);
  });
  test('snapshot', () => {
    expect(el).toMatchSnapshot();
  });

  describe('component', () => {
    test('rendering', () => {
      expect(el.isEmptyRender()).toEqual(false);
      const optionsEl = el.find('.criteria-option');
      expect(optionsEl.length).toEqual(props.config.options.length);
      optionsEl.forEach((optionEl, i) => {
        let option = props.config.options[i];
        expect(optionEl.key()).toEqual(option.name);
        expect(optionEl.find('.option-label').childAt(0).text()).toEqual(
          option.label,
        );
        expect(optionEl.find('.option-points').childAt(0).text()).toContain(
          String(option.points),
        );
      });
    });
  });

  describe('mapStateToProps', () => {
    const testState = { arbitary: 'some data' };
    const ownProps = { orderNum: props.orderNum };
    let mapped;
    beforeEach(() => {
      mapped = mapStateToProps(testState, ownProps);
    });
    test('selectors.app.rubric.criterionConfig', () => {
      expect(mapped.config).toEqual(
        selectors.app.rubric.criterionConfig(testState, ownProps),
      );
    });

    test('selectors.grading.selected.criterionGradeData', () => {
      expect(mapped.data).toEqual(
        selectors.grading.selected.criterionGradeData(testState, ownProps),
      );
    });
  });
});
