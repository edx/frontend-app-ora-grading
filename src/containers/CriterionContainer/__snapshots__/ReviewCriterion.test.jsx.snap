// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Review Crition Container snapshot 1`] = `undefined`;
